File: DemoApp Objective-C/CASSample/ViewController.m
//
//  ViewController.m
//  CASSample
//

#import "ViewController.h"
#import "AppDelegate.h"

@interface ViewController ()

@end

@implementation ViewController

BOOL isAppReturnEnable = false;

- (void)viewDidLoad {
	[super viewDidLoad];
	self.bannerDelegate = [[BannerAdDelegate alloc] init];
    self.bannerDelegate.infoLabel = self.statusBannerLabel;
	self.interDelegate = [[AdDelegate alloc] initWithType:CASTypeInterstitial];
	self.rewardDelegate = [[AdDelegate alloc] initWithType:CASTypeRewarded];


	[self.versionLabel setText:[CAS getSDKVersion]];

	CASMediationManager *manager = [AppDelegate getMediationManager];
	[manager setAdLoadDelegate:self];

	// Create Banner view from Code
	//[self createBannerView];
	// Else use Banner view from Storyboard

	[self.bannerView setTranslatesAutoresizingMaskIntoConstraints:NO];
	[self.bannerView setAdSize:[CASSize getSmartBanner]];
	[self.bannerView setRootViewController:self];
	[self.bannerView setAdDelegate:self.bannerDelegate];

	self.statusBannerLabel.text = _bannerView.isAdReady ? @"Ready" : @"Loading";
	self.statusInterstitialLabel.text = manager.isInterstitialReady ? @"Ready" : @"Loading";
	self.statusRewardedLabel.text = manager.isRewardedAdReady ? @"Ready" : @"Loading";
}

-(void)createBannerView {
	CASSize *size = [CASSize getSmartBanner];
	CASMediationManager* manager = [AppDelegate getMediationManager];
	self.bannerView = [[CASBannerView alloc] initWithAdSize:size manager:manager];

	[self.view addSubview:self.bannerView];

	if (@available(iOS 11.0, *)) {
		UILayoutGuide *guide = self.view.safeAreaLayoutGuide;
		[NSLayoutConstraint activateConstraints:@[
			 [NSLayoutConstraint constraintWithItem:self.bannerView
			  attribute:NSLayoutAttributeBottom
			  relatedBy:NSLayoutRelationEqual
			  toItem:guide
			  attribute:NSLayoutAttributeBottom
			  multiplier:1
			  constant:0],
			 [NSLayoutConstraint constraintWithItem:self.bannerView
			  attribute:NSLayoutAttributeRight
			  relatedBy:NSLayoutRelationEqual
			  toItem:guide
			  attribute:NSLayoutAttributeRight
			  multiplier:1
			  constant:0]
		]];
	} else {
		[NSLayoutConstraint activateConstraints:@[
			 [NSLayoutConstraint constraintWithItem:self.bannerView
			  attribute:NSLayoutAttributeBottom
			  relatedBy:NSLayoutRelationEqual
			  toItem:self.view
			  attribute:NSLayoutAttributeBottom
			  multiplier:1
			  constant:0],
			 [NSLayoutConstraint constraintWithItem:self.bannerView
			  attribute:NSLayoutAttributeRight
			  relatedBy:NSLayoutRelationEqual
			  toItem:self.view
			  attribute:NSLayoutAttributeRight
			  multiplier:1
			  constant:0]
		]];
	}
}

- (IBAction)showBanner:(id)sender {
	[self.bannerView setHidden:NO];
}

- (IBAction)hideBanner:(id)sender {
	[self.bannerView setHidden:YES];
}

- (IBAction)setBannerSize:(id)sender {
	[self.bannerView setAdSize:CASSize.banner];
}

- (IBAction)setLeaderboardSize:(id)sender {
	[self.bannerView setAdSize:CASSize.leaderboard];
}

- (IBAction)setMrecSize:(id)sender {
	[self.bannerView setAdSize:CASSize.mediumRectangle];
}

- (IBAction)setAdaptiveSize:(id)sender {
	[self.bannerView setAdSize:[CASSize getAdaptiveBannerInContainer:self.view]];
}

- (IBAction)setSmartSize:(id)sender {
	[self.bannerView setAdSize:[CASSize getSmartBanner]];
}

- (IBAction)showInterstitial:(id)sender {
	CASMediationManager* manager = [AppDelegate getMediationManager];
	[manager presentInterstitialFromRootViewController:self callback:self.interDelegate];
}

- (IBAction)showRewarded:(id)sender {
	CASMediationManager* manager = [AppDelegate getMediationManager];
	[manager presentRewardedAdFromRootViewController:self callback:self.rewardDelegate];
}

- (IBAction)changeStateOfAppReturn {
	if (isAppReturnEnable) {
		_appReturnStatusLabel.text = @"Disabled";
		[_changeStateOfAppReturnButton setTitle:@"Enable" forState:UIControlStateNormal];
		[CAS.manager disableAppReturnAds];
		isAppReturnEnable = false;
	} else {
		_appReturnStatusLabel.text = @"Enabled";
		[_changeStateOfAppReturnButton setTitle:@"Disable" forState:UIControlStateNormal];
		[CAS.manager enableAppReturnAdsWith:self];
		isAppReturnEnable = true;
	}
}

#pragma mark CASLoadDelegate implementation
- (void)onAdLoaded:(enum CASType)adType {
	// CASLoadDelegate called from background thread. To use UI API we switch to main thread.
	dispatch_async(dispatch_get_main_queue(), ^{
		switch (adType) {
		case CASTypeInterstitial:
			NSLog(@"[CAS Sample] Interstitial Ad loaded and ready to present");
			self.statusInterstitialLabel.text = @"Ready";
			break;
		case CASTypeRewarded:
			NSLog(@"[CAS Sample] Rewarded Ad loaded and ready to present");
			self.statusRewardedLabel.text = @"Ready";
			break;
		case CASTypeBanner:
			// CASLoadDelegate protocol should be used to listen Interstitial and Rewarded Ads only.
			// Listen Banner ads by CASBannerDelegate protocol.
			break;
		default:
			break;
		}
	});
}

- (void)onAdFailedToLoad:(enum CASType) adType withError:(NSString *)error {
	// CASLoadDelegate called from background thread. To use UI API we switch to main thread.
	dispatch_async(dispatch_get_main_queue(), ^{
		switch (adType) {
		case CASTypeInterstitial:
			NSLog(@"[CAS Sample] Interstitial Ad failed to load with error: %@", error);
			self.statusInterstitialLabel.text = @"Ready";
			break;
		case CASTypeRewarded:
			NSLog(@"[CAS Sample] Rewarded Ad failed to load with error: %@", error);
			self.statusRewardedLabel.text = @"Ready";
			break;
		case CASTypeBanner:
			// CASLoadDelegate protocol should be used to listen Interstitial and Rewarded Ads only.
			// Listen Banner ads by CASBannerDelegate protocol.
			break;
		default:
			break;
		}
	});
}

#pragma mark CASAppReturnDelegate implementation
- (UIViewController *)viewControllerForPresentingAppReturnAd {
	return self;
}

@end

File: Swift/CASSample/ViewController.swift
//
//  ViewController.swift
//  CASSample
//
//  Copyright © 2022 Clever Ads Solutions. All rights reserved.
//

import CleverAdsSolutions
import UIKit

class ViewController: UIViewController, CASLoadDelegate, CASAppReturnDelegate {
    @IBOutlet var versionLabel: UILabel!
    @IBOutlet var bannerView: CASBannerView!

    @IBOutlet var statusBannerLabel: UILabel!
    @IBOutlet var statusInterstitialLabel: UILabel!
    @IBOutlet var statusRewardedLabel: UILabel!
    @IBOutlet var statusAppReturnLabel: UILabel!
    @IBOutlet var enableAppReturnButton: UIButton!

    private var manager: CASMediationManager!

    private let bannerDelegate = BannerAdDelegate()
    private let interDelegate = AdContentDelegate(type: .interstitial)
    private let rewardDelegate = AdContentDelegate(type: .rewarded)

    private var isAppReturnEnabled: Bool = false

    override func viewDidLoad() {
        super.viewDidLoad()
        versionLabel.text = CAS.getSDKVersion()

        // Get last initialized CAS Manager.
        let lastManager = AppDelegate.mediationManager!
        manager = lastManager
        lastManager.adLoadDelegate = self

        // CASBannerView created in Storyboard else use createBannerAdView() to create
        // createBannerAdView(lastManager)

        bannerView.translatesAutoresizingMaskIntoConstraints = false
        bannerView.adSize = CASSize.getSmartBanner()
        bannerView.adDelegate = bannerDelegate
        bannerView.rootViewController = self

        bannerDelegate.infoLabel = statusBannerLabel

        statusBannerLabel.text = bannerView.isAdReady ? "Ready" : "Laoding"
        statusInterstitialLabel.text = lastManager.isInterstitialReady ? "Ready" : "Laoding"
        statusRewardedLabel.text = lastManager.isRewardedAdReady ? "Ready" : "Laoding"
    }

    func createBannerAdView(_ manager: CASMediationManager) {
        bannerView = CASBannerView(adSize: CASSize.getSmartBanner(), manager: manager)
        view.addSubview(bannerView)

        if #available(iOS 11, *) {
            let guide = view.safeAreaLayoutGuide
            NSLayoutConstraint.activate([
                bannerView.bottomAnchor.constraint(equalTo: guide.bottomAnchor),
                bannerView.rightAnchor.constraint(equalTo: guide.rightAnchor),
            ])
        } else {
            NSLayoutConstraint.activate([
                bannerView.bottomAnchor.constraint(equalTo: view.bottomAnchor),
                bannerView.rightAnchor.constraint(equalTo: view.rightAnchor),
            ])
        }
    }

    @IBAction func showBanner(_ sender: Any) {
        bannerView.isHidden = false
    }

    @IBAction func hideBanner(_ sender: Any) {
        bannerView.isHidden = true
    }

    @IBAction func setBannerSize(_ sender: Any) {
        bannerView.adSize = .banner
    }

    @IBAction func setLeaderboardSize(_ sender: Any) {
        bannerView.adSize = .leaderboard
    }

    @IBAction func setMrecSize(_ sender: Any) {
        bannerView.adSize = .mediumRectangle
    }

    @IBAction func setAdaptiveSize(_ sender: Any) {
        bannerView.adSize = .getAdaptiveBanner(inContainer: view)
    }

    @IBAction func setSmartSize(_ sender: Any) {
        bannerView.adSize = .getSmartBanner()
    }

    @IBAction func showInterstitial(_ sender: Any) {
        manager.presentInterstitial(fromRootViewController: self,
                                    callback: interDelegate)
    }

    @IBAction func showRewarded(_ sender: Any) {
        manager.presentRewardedAd(fromRootViewController: self,
                                  callback: rewardDelegate)
    }

    @IBAction func changeStateOfAppReturn(_ sender: Any) {
        if isAppReturnEnabled {
            enableAppReturnButton.setTitle("Enable", for: UIControl.State.normal)
            statusAppReturnLabel.text = "Disabled"
            manager?.disableAppReturnAds()
            isAppReturnEnabled = false
        } else {
            enableAppReturnButton.setTitle("Disable", for: UIControl.State.normal)
            statusAppReturnLabel.text = "Enabled"
            manager?.enableAppReturnAds(with: self)
            isAppReturnEnabled = true
        }
    }

    // MARK: CASLoadDelegate implementation

    func onAdLoaded(_ adType: CASType) {
        // CASLoadDelegate called from background thread. To use UI API we switch to main thread.
        DispatchQueue.main.async {
            switch adType {
            case .interstitial:
                print("[CAS Sample] Interstitial Ad loaded and ready to present")
                self.statusInterstitialLabel.text = "Ready"
            case .rewarded:
                print("[CAS Sample] Rewarded Ad loaded and ready to present")
                self.statusRewardedLabel.text = "Ready"
            case .banner:
                // CASLoadDelegate protocol should be used to listen Interstitial and Rewarded Ads only.
                // Listen Banner ads by CASBannerDelegate protocol.
                break
            default: break
            }
        }
    }

    func onAdFailedToLoad(_ adType: CASType, withError error: String?) {
        // CASLoadDelegate called from background thread. To use UI API we switch to main thread.
        DispatchQueue.main.async {
            switch adType {
            case .interstitial:
                print("[CAS Sample] Interstitial Ad failed to load with error: \(String(describing: error))")
                self.statusInterstitialLabel.text = error ?? "Failed"
            case .rewarded:
                print("[CAS Sample] Rewarded Ad failed to load with error: \(String(describing: error))")
                self.statusRewardedLabel.text = error ?? "Failed"
            case .banner:
                // CASLoadDelegate protocol should be used to listen Interstitial and Rewarded Ads only.
                // Listen Banner ads by CASBannerDelegate protocol.
                break
            default: break
            }
        }
    }

    // MARK: CASAppReturnDelegate implementation

    // Implement the method to pass active UIViewController
    func viewControllerForPresentingAppReturnAd() -> UIViewController {
        return self
    }
}
